src/main/java/gtPlusPlus/xmod/gregtech/common/tileentities/machines/multi/production/GregtechMetaTileEntity_ExplosionGenerator.java

```diff
@@ -0,0 +1,288 @@
+    }
+
+    @Override
+    protected int getCasingTextureId() {
+        return 11;
+    }
+
+    @Override
+    public boolean isCorrectMachinePart(ItemStack aStack) {
+        return getMaxEfficiency(aStack) > 0;
+    }
+
+    @Override
+    public boolean checkRecipe(final ItemStack aStack) {
+        this.mProgresstime = 0;
+        this.mMaxProgresstime = 200;
+        this.lEUt = 0;
+        this.mEfficiencyIncrease = 10000;
+        return true;
+    }
+
+    @Override
+    public IStructureDefinition<GregtechMetaTileEntity_ExplosionGenerator> getStructureDefinition() {
+        if (STRUCTURE_DEFINITION == null) {
+            STRUCTURE_DEFINITION = StructureDefinition.<GregtechMetaTileEntity_ExplosionGenerator>builder().addShape(
+                    mName,
+                    transpose(
+                            new String[][] { { "HHH", "HCH", "HCH", "   ", "SSS" },
+                                    { "H~H", "CGC", "CGC", " S ", "SSS" }, { "HHH", "HCH", "HCH", "   ", "SSS" }, }))
+                    // Regular casings or hatches
+                    .addElement(
+                            'C',
+                            buildHatchAdder(GregtechMetaTileEntity_ExplosionGenerator.class)
+                                    .atLeast(ImmutableMap.of(Maintenance, 1, Dynamo, 1)).casingIndex(11).dot(1)
+                                    .buildAndChain(
+                                            onElementPass(x -> ++x.mCasing, ofBlock(GregTech_API.sBlockCasings1, 11))))
+                    // Gearbox
+                    .addElement('G', ofBlock(GregTech_API.sBlockCasings2, 3))
+                    // Heat-Proof Casings
+                    .addElement('H', ofBlock(GregTech_API.sBlockCasings1, 11))
+                    // Solid Steel
+                    .addElement('S', ofBlock(GregTech_API.sBlockCasings2, 0)).build();
+        }
+        return STRUCTURE_DEFINITION;
+    }
+
+    @Override
+    public boolean onRunningTick(ItemStack aStack) {
+        mProgresstime = 0;
+        IGregTechTileEntity baseTile = getBaseMetaTileEntity();
+        int offsetX = ForgeDirection.getOrientation(baseTile.getBackFacing()).offsetX * -6;
+        int offsetY = ForgeDirection.getOrientation(baseTile.getBackFacing()).offsetY * -6;
+        int offsetZ = ForgeDirection.getOrientation(baseTile.getBackFacing()).offsetZ * -6;
+        List<Entity> entities = baseTile.getWorld().getEntitiesWithinAABBExcludingEntity(
+                null,
+                AxisAlignedBB.getBoundingBox(
+                        baseTile.getXCoord() - 0.5 - offsetX,
+                        baseTile.getYCoord() - 0.5 - offsetY,
+                        baseTile.getZCoord() - 0.5 - offsetZ,
+                        baseTile.getXCoord() + 1.5 - offsetX,
+                        baseTile.getYCoord() + 1.5 - offsetY,
+                        baseTile.getZCoord() + 1.5 - offsetZ));
+        for (Entity tnt : entities) {
```

Comment:
I would suggest listening for `net.minecraft.world.Explosion` event. Running a get entity within aabb call per tick is slightly concerning to me, performance wise. would also remove the need for sending sound events manually

Reply:
yeah I agree a machine idling at 10k ns isn't ideal but wouldn't the explosion have to happen to listen for it? this is taken from the botania flower that calls aabb in a 24x24 area every tick.